File: reporting/user_integrity.py

from __future__ import absolute_import

import os
import urllib
import unicodedata
import logging
import base64
import hashlib
import json
import re
from datetime import tzinfo, timedelta, datetime

from google.appengine.api import users
from google.appengine.ext import ndb
from google.appengine.api import app_identity
from google.appengine.api import mail
from google.appengine.api import images
from google.appengine.ext import blobstore

import jinja2
import webapp2

import cloudstorage as gcs

import constants
from pyutils import dt_utils
from pyutils import utils, Utils
from reporting import reporting_utils
from db import ControlParameters

# import gc

class Integrity(webapp2.RequestHandler):
    KEY = 'integrity'

    def get(self):
        try:
            is_cron = self.request.headers['X-Appengine-Cron']
        except Exception as e:
            is_cron = False

        user = users.get_current_user()
        if user:
            user_email_addr = user.email()
        else:
            user_email_addr = ""

        if not is_cron and user_email_addr not in constants.LIST_OF_ADMINS:
            self.response.write("<b>UN-AUTHORIZED</b>")
        else:
            if self.request.path == '/integrity/user-integrity/get-by-user':
                self.response.write(self.get_user_integrity_by_user())
            elif self.request.path in ('/integrity/user-integrity/load', '/jobs/integrity/user-integrity/load'):
                self.load_user_integrity()
            elif self.request.path in ('/integrity/user-integrity/postload', '/jobs/integrity/user-integrity/postload'):
                self.post_load_user_integrity()

    def post(self):
        user = users.get_current_user()
        if user:
            user_email_addr = user.email()
        else:
            user_email_addr = ""

        if user_email_addr not in constants.LIST_OF_ADMINS:
            self.response.write("<b>UN-AUTHORIZED</b>")
        else:
            if self.request.path in ('/integrity/user-integrity/load', '/jobs/integrity/user-integrity/load'):
                self.load_user_integrity()
            elif self.request.path in ('/integrity/user-integrity/postload', '/jobs/integrity/user-integrity/postload'):
                self.post_load_user_integrity()

    def get_user_integrity_by_user(self):
        _f = gcs.open(constants.USER_INTEGRITY_BY_USER)
        _contents = _f.read()
        _f.close()
        return _contents

    def load_user_integrity(self):
        reporting_fields = {
            'i_fname': '',
            'i_lname': '',
            'i_enrolled_people': '',
            'i_org_courses': [
                'i_org_course_from_date',
                'i_org_course_to_date',
                'i_org_course_leadteacher',
                'i_org_course_city',
                'i_org_course_state',
                'i_org_course_leadteacher',
            ],
        }

        ttc_list = reporting_utils.get_ttc_list()

        min_updated_datetime = None
        cparams = ControlParameters.fetch()
        if cparams:
            min_updated_datetime = cparams.user_integrity_last_update_datetime
            if min_updated_datetime:
                logging.info('[load_user_integrity] user_integrity_last_update_datetime = ' + min_updated_datetime.strftime('%Y-%m-%d %H:%M:%S'))
            else:
                logging.info('[load_user_integrity] user_integrity_last_update_datetime = NONE')
        try:
            _f = gcs.stat(constants.USER_INTEGRITY_BY_USER)
            file_updated_datetime = datetime.fromtimestamp(_f.st_ctime)
            if not min_updated_datetime:
                min_updated_datetime = (file_updated_datetime - timedelta(hours=1))
                logging.error('[load_user_integrity] Did not find param for "user_integrity_last_update_datetime". Setting default from update date of ' + constants.USER_INTEGRITY_BY_USER)
            else:
                if file_updated_datetime < min_updated_datetime:
                    min_updated_datetime = (file_updated_datetime - timedelta(hours=1))
            _f = gcs.open(constants.USER_INTEGRITY_BY_USER)
            _user_data_by_email = json.loads(_f.read())
            _f.close()
        except gcs.NotFoundError:
            logging.info('[load_user_integrity] Failed to load the data from old user_integrity')
            logging.info('[load_user_integrity] Initializing user with empty dictionary')
            _user_data_by_email = {}
            min_updated_datetime = None

        if min_updated_datetime:
            logging.info('[load_user_integrity] user_integrity_last_update_datetime = ' + min_updated_datetime.strftime('%Y-%m-%d %H:%M:%S'))

        current_datetime = datetime.now()
        _user_files = Utils.list_files(
            # prefix=constants.USER_CONFIG_FOLDER,
            prefix=constants.USER_CONFIG_LOCATION,
            min_updated_datetime=min_updated_datetime,
        )

        # _user_files = gcs.listbucket(constants.USER_CONFIG_LOCATION)
        # _user_data_by_email = {}
        for _f in _user_files:
            if _f.filename.endswith('.json') and not '/summary/' in _f.filename and not '/integrity/' in _f.filename:
                # logging.info('[load_user_integrity] Reading file ' + _f.filename)
                gcs_file = gcs.open(_f.filename)
                _contents = gcs_file.read()
                if _contents and _contents.strip() != '':
                    _ud = json.loads(_contents)
                    _ue = _ud.get('email', '')
                    gcs_file.close()
                    if 'form_data' in _ud:
                        for _ft in _ud['form_data']:
                            if _ft != 'ttc_application':
                                continue
                            for _fi_raw in _ud['form_data'][_ft]:
                                if _fi_raw != 'default':
                                    _fd = _ud['form_data'][_ft][_fi_raw]
                                    _fd['form_instance'] = _fi_raw
                                    if 'last_update_datetime' in _fd:
                                        _fd['last_update_datetime_est'] = dt_utils.utc_to_timezone(
                                            datetime.strptime(
                                                _fd['last_update_datetime'], 
                                                "%Y-%m-%d %H:%M:%S"
                                            ), 
                                            dt_utils.Eastern
                                        ).strftime("%Y-%m-%d %H:%M:%S")

                                    if _fd['last_update_datetime_est'] < datetime.strftime(datetime.now() - timedelta(constants.DATA_RETENTION_DAYS), "%Y-%m-%d %H:%M:%S"):
                                        continue

                                    # [START] [reporting_fields] Processing application status
                                    app_status, eval_status = reporting_utils.get_reporting_status(
                                        _ft, 
                                        _fd.get('is_form_submitted', False),
                                        _fd.get('is_form_complete', False),
                                    )
                                    # [END] [reporting_fields] Processing application status

                                    # [START] [photo_url] Setting photo public URL
                                    # [END] [photo_url] Setting photo public URL

                                    # [START] [reporting_fields] Processing form_instance_page_data
                                    if 'form_instance_page_data' in _fd and 'i_ttc_country_and_dates' in _fd['form_instance_page_data']:
                                        _fd['ttc_metadata'] = ttc_list.get(_fd['form_instance_page_data']['i_ttc_country_and_dates'], {})
                                    # [END] [reporting_fields] Processing form_instance_page_data

                                    # [START] [reporting_fields] Processing all answers
                                    if 'data' in _fd:
                                        _fd_data = {}
                                        for _q, _qa in _fd['data'].items():
                                            # Storing only whitelisted fields
                                            if _q in reporting_fields:
                                                _qa2 = _qa
                                                if isinstance(_qa, dict) and isinstance(reporting_fields[_q], list):
                                                    _qa2 = {}
                                                    for _tq, _tqa in _qa.items():
                                                        if _tq in reporting_fields[_q]:
                                                            _qa2[_tq] = _tqa
                                                _fd_data[_q] = _qa2

                                        _fd['data'] = _fd_data
                                    # [END] [reporting_fields] Processing all answers

                                    _fi_email = None
                                    if '-' in _fi_raw:
                                        _fi_arr = _fi_raw.split('-')
                                        # Checking against 2nd value, since sometimes people enter names instead of emails
                                        # That can throw off this logic. So in such cases we just take the first value as the email.
                                        if '@' not in _fi_arr[1]:
                                            _fi_email = _fi_arr[0]
                                            _fi = _fi_arr[1]
                                        else:
                                            _fi = _fi_arr[0]
                                            _fi_email = _fi_arr[1]
                                    else:
                                        _fi = _fi_raw

                                    if _ue not in _user_data_by_email:
                                        _user_data_by_email[_ue] = {}
                                    if _ft not in _user_data_by_email[_ue]:
                                        _user_data_by_email[_ue][_ft] = {}
                                        _user_data_by_email[_ue][_ft][Integrity.KEY] = {}
                                    if _fi not in _user_data_by_email[_ue][_ft]:
                                        _user_data_by_email[_ue][_ft][_fi] = {}

                                    if _fi_email:
                                        _user_data_by_email[_ue][_ft][_fi][_fi_email] = _fd
                                    else:
                                        _user_data_by_email[_ue][_ft][_fi] = _fd


        # [START KEYRESET] Clearing past evaluation assignments
        for _e in _user_data_by_email:
            if 'ttc_application' in _user_data_by_email[_e]:
                _user_data_by_email[_e]['ttc_application'][Integrity.KEY]['enrolled_matches'] = {}
                _user_data_by_email[_e]['ttc_application'][Integrity.KEY]['org_course_matches'] = {}
        # [END KEYRESET]

        # logging.info('[load_user_integrity] Processing data')

        # logging.info('[load_user_integrity] Processing applications')
        for _c1e in _user_data_by_email:
            for _c1fi in _user_data_by_email[_c1e].get('ttc_application', {}):
                if _c1fi == 'default':
                    continue
                # logging.info('[load_user_integrity] Processing application for ' + _c1e)
                _c1 = _user_data_by_email[_c1e]['ttc_application'][_c1fi]
                _c1d = _c1.get('data', {})
                _c1fn = _c1d.get('i_fname','').strip().lower()
                _c1ln = _c1d.get('i_lname','').strip().lower()
                _c1n = _c1fn + " " + _c1ln

                for _c2e in _user_data_by_email:
                    if _c1e == _c2e:
                        continue
                    # logging.info('[load_user_integrity][' + _c1e + '] Comparing against application for: ' + _c2e)
                    for _c2fi in _user_data_by_email[_c2e].get('ttc_application', {}):
                        _c2 = _user_data_by_email[_c2e]['ttc_application'][_c2fi]
                        _c2d = _c2.get('data', {})
                        _c2fn = _c2d.get('i_fname','').strip().lower()
                        _c2ln = _c2d.get('i_lname','').strip().lower()
                        _c2n = _c2fn + " " + _c2ln

                        if _c1n == _c2n:
                            continue

                        for _c1en in _c1d.get('i_enrolled_people', {}):
                            for _c2en in _c2d.get('i_enrolled_people', {}):
                                is_matched = False
                                _ee1 = _c1en['i_enrollment_email'].strip().lower()
                                _ee2 = _c2en['i_enrollment_email'].strip().lower()
                                _en1 = _c1en['i_enrollment_name'].strip().lower()
                                _en2 = _c2en['i_enrollment_name'].strip().lower()
                                _ec1 = _c1en['i_enrollment_city'].strip().lower()
                                _ec2 = _c2en['i_enrollment_city'].strip().lower()
                                _es1 = _c1en['i_enrollment_state'].strip().lower()
                                _es2 = _c2en['i_enrollment_state'].strip().lower()
                                if '@' in _ee1 and _ee1 == _ee2:
                                    is_matched = True
                                if _en1 == _en2 and _ec1 == _ec2 and _es1 == _es2:
                                    is_matched = True

                                if is_matched:
                                    if Integrity.KEY not in _user_data_by_email[_c1e]['ttc_application']:
                                        _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY] = {}
                                    if 'enrolled_matches' not in _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]:
                                        _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['enrolled_matches'] = {}
                                    if _c2e not in _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['enrolled_matches']:
                                        _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['enrolled_matches'][_c2e] = set()
                                    _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['enrolled_matches'][_c2e].add("{_en2} <{_ee2}>".format(**locals()))

                        for _c1en in _c1d.get('i_org_courses', {}):
                            _ofd1 = _c1en['i_org_course_from_date'].strip()
                            _otd1 = _c1en['i_org_course_to_date'].strip()
                            _ot1 = _c1en['i_org_course_leadteacher'].strip().lower()
                            _oc1 = _c1en['i_org_course_city'].strip().lower()
                            _os1 = _c1en['i_org_course_state'].strip().lower()

                            _olt1 = _c1en["i_org_course_leadteacher"].strip().lower()
                            _re = re.search("^[^a-z]{0,3}([a-z]+ [a-z]+)", _olt1, re.IGNORECASE)
                            if _re:
                                _oltn1 = _re.group(1)
                            else:
                                _oltn1 = ''

                            for _c2en in _c2d.get('i_org_courses', {}):
                                is_matched = False
                                _ofd2 = _c2en['i_org_course_from_date'].strip()
                                _otd2 = _c2en['i_org_course_to_date'].strip()
                                _ot2 = _c2en['i_org_course_leadteacher'].strip().lower()
                                _oc2 = _c2en['i_org_course_city'].strip().lower()
                                _os2 = _c2en['i_org_course_state'].strip().lower()

                                _olt2 = _c2en["i_org_course_leadteacher"].strip().lower()
                                _re = re.search("^[^a-z]{0,3}([a-z]+ [a-z]+)", _olt2, re.IGNORECASE)
                                if _re:
                                    _oltn2 = _re.group(1)
                                else:
                                    _oltn2 = ''

                                if (
                                    # _oc1 != 'online' and
                                    # _os1 != 'online' and
                                    (_ofd1 == _ofd2 or _otd1 == _otd2) and 
                                    _oc1 == _oc2 and 
                                    _os1 == _os2 and
                                    _oltn1 and
                                    _oltn2 and
                                    (
                                        _oltn1 == _oltn2 or
                                        _oltn1 in _olt2 or
                                        _oltn2 in _olt1
                                    )
                                ):
                                    is_matched = True

                                if is_matched:
                                    if Integrity.KEY not in _user_data_by_email[_c1e]['ttc_application']:
                                        _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY] = {}
                                    if 'org_course_matches' not in _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]:
                                        _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['org_course_matches'] = {}
                                    if _c2e not in _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['org_course_matches']:
                                        _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['org_course_matches'][_c2e] = set()
                                    _user_data_by_email[_c1e]['ttc_application'][Integrity.KEY]['org_course_matches'][_c2e].add("{_ofd2} - {_otd2} ({_oc2}, {_os2} - {_oltn2})".format(**locals()))

        # ----------------------------------------------------------------------------------------
        write_retry_params = gcs.RetryParams(backoff_factor=1.1)
        _f = gcs.open(
            constants.USER_INTEGRITY_BY_USER,
            'w',
            content_type='text/plain',
            options={'x-goog-meta-foo': 'foo',
            'x-goog-meta-bar': 'bar'},
            retry_params=write_retry_params
        )
        _f.write(json.dumps(_user_data_by_email, default=utils.json_dumps_set_default))
        _f.close()

        ControlParameters.create(
            {
                'user_integrity_last_update_datetime': current_datetime,
            }
        )
    # ----------------------------------------------------------------------------------------

    def post_load_user_integrity(self):
        try:
            _f = gcs.open(constants.USER_INTEGRITY_BY_USER)
            _user_data_by_email = json.loads(_f.read())
            _f.close()
        except gcs.NotFoundError:
            logging.info('[post_load_user_integrity] Failed to load the data from old user_integrity')
            logging.info('[post_load_user_integrity] Initializing user with empty dictionary')
            _user_data_by_email = {}

        enrolledCSV = "Applicant Name,Applicant Email,Enrolled Name,Enrolled Email\n"
        for _c1e in _user_data_by_email:
            for _c1fi in _user_data_by_email[_c1e].get('ttc_application', {}):
                if _c1fi == 'default':
                    continue
                logging.info('[post_load_user_integrity] Processing application for ' + _c1e)
                _c1 = _user_data_by_email[_c1e]['ttc_application'][_c1fi]
                _c1d = _c1.get('data', {})
                _c1fn = _c1d.get('i_fname','').strip().lower()
                _c1ln = _c1d.get('i_lname','').strip().lower()
                _c1n = _c1fn + " " + _c1ln
                enrolledCSV += u"{_c1n},{_c1e},,\n".format(**locals())
                for _c1en in _c1d.get('i_enrolled_people', {}):
                    _ee1 = _c1en['i_enrollment_email'].strip().lower()
                    _en1 = _c1en['i_enrollment_name'].strip().lower()
                    enrolledCSV += u",,{_en1},{_ee1}\n".format(**locals())

        write_retry_params = gcs.RetryParams(backoff_factor=1.1)
        _fecsv = gcs.open(
            constants.APPLICANT_ENROLLED_LIST,
            'w',
            content_type='text/plain',
            options={'x-goog-meta-foo': 'foo',
            'x-goog-meta-bar': 'bar'},
            retry_params=write_retry_params
        )
        import unicodedata
        enrolledCSV = unicodedata.normalize('NFKD', enrolledCSV).encode('ascii', 'ignore')
        _fecsv.write(enrolledCSV)
        # logging.info('[load_user_integrity] Processing applications')
        _fecsv.close()
    # ----------------------------------------------------------------------------------------

#[END Integrity]

app = webapp2.WSGIApplication(
    [
        ('/integrity/user-integrity/load', Integrity),
        ('/integrity/user-integrity/get-by-user', Integrity),
        ('/jobs/integrity/user-integrity/load', Integrity),
        ('/jobs/integrity/user-integrity/postload', Integrity),
    ],
    debug=True
)

File: app/utils/reporting/user-integrity.ts
import { readJson, readText, writeJson, writeText, listFiles, GCS_PATHS } from '../gcs';
import { getTtcList, getReportingStatus } from './reporting-utils';
import { getReportableInstanceKeys } from './form-instance-normalizer';

const DATA_RETENTION_DAYS = 730;
const KEY = 'integrity';

/** Convert UTC datetime string to Eastern timezone display string. */
function utcToEastern(utcStr: string): string {
  const d = new Date(utcStr + ' UTC');
  return d.toLocaleString('en-US', {
    timeZone: 'America/New_York',
    year: 'numeric', month: '2-digit', day: '2-digit',
    hour: '2-digit', minute: '2-digit', second: '2-digit',
    hour12: false,
  }).replace(/(\d+)\/(\d+)\/(\d+),?\s*/, '$3-$1-$2 ');
}

/* eslint-disable @typescript-eslint/no-explicit-any */

/** Recursively convert Set instances to Arrays for JSON serialization. */
function convertSetsToArrays(obj: any): void {
  if (!obj || typeof obj !== 'object') return;
  if (obj instanceof Set) return; // handled by parent
  if (Array.isArray(obj)) {
    obj.forEach(convertSetsToArrays);
  } else {
    for (const key of Object.keys(obj)) {
      if (obj[key] instanceof Set) {
        obj[key] = Array.from(obj[key]);
      } else {
        convertSetsToArrays(obj[key]);
      }
    }
  }
}

/**
 * Port of Integrity.load_user_integrity() (Python lines 80-357).
 * Reads all user data, compares enrolled people and organized courses
 * across applicants to find duplicates/matches.
 */
export async function loadUserIntegrity(): Promise<void> {
  const reportingFields: Record<string, string | string[]> = {
    'i_fname': '',
    'i_lname': '',
    'i_enrolled_people': '',
    'i_org_courses': [
      'i_org_course_from_date',
      'i_org_course_to_date',
      'i_org_course_leadteacher',
      'i_org_course_city',
      'i_org_course_state',
      'i_org_course_leadteacher',
    ],
  };

  const ttcList = await getTtcList();

  // Load existing integrity data or start fresh
  let userDataByEmail: any = {};
  try {
    userDataByEmail = await readJson(GCS_PATHS.USER_INTEGRITY_BY_USER);
  } catch {
    // File doesn't exist yet
  }

  // List all user JSON files
  const userFiles = await listFiles(GCS_PATHS.USER_CONFIG_PREFIX);
  const now = new Date();

  for (const file of userFiles) {
    if (!file.name.endsWith('.json') || file.name.includes('/summary/') || file.name.includes('/integrity/')) {
      continue;
    }

    let contents: string;
    try {
      contents = await readText(file.name);
    } catch {
      continue;
    }

    if (!contents || contents.trim() === '') continue;

    let ud: any;
    try {
      ud = JSON.parse(contents);
    } catch {
      continue;
    }

    const ue: string = ud.email || '';
    if (!ue) continue;
    if (!ud.form_data) continue;

    for (const ft of Object.keys(ud.form_data)) {
      if (ft !== 'ttc_application') continue;

      for (const fiRaw of getReportableInstanceKeys(ud.form_data[ft])) {

        const fd: any = { ...ud.form_data[ft][fiRaw] };
        fd.form_instance = fiRaw;

        // Convert last_update_datetime to Eastern (matching user-summary.ts)
        if (fd.last_update_datetime) {
          fd.last_update_datetime_est = utcToEastern(fd.last_update_datetime);
        }

        // Check retention (skip old records) — port of Python line 161
        if (fd.last_update_datetime) {
          const cutoffDate = new Date(now);
          cutoffDate.setDate(cutoffDate.getDate() - DATA_RETENTION_DAYS);
          const lastUpdate = new Date(fd.last_update_datetime);
          if (lastUpdate < cutoffDate) continue;
        }

        // Reporting status
        const [appStatus, evalStatus] = getReportingStatus(
          ft,
          fd.is_form_submitted || false,
          fd.is_form_complete || false,
        );
        fd.app_status = appStatus;
        fd.eval_status = evalStatus;

        // TTC metadata
        if (fd.form_instance_page_data?.i_ttc_country_and_dates) {
          fd.ttc_metadata = ttcList[fd.form_instance_page_data.i_ttc_country_and_dates] || {};
        }

        // Filter data to whitelisted fields only
        if (fd.data) {
          const fdData: any = {};
          for (const [q, qa] of Object.entries(fd.data)) {
            if (q in reportingFields) {
              let qa2 = qa;
              const fieldSpec = reportingFields[q];
              if (qa && typeof qa === 'object' && !Array.isArray(qa) && Array.isArray(fieldSpec)) {
                qa2 = {};
                for (const [tq, tqa] of Object.entries(qa as any)) {
                  if (fieldSpec.includes(tq)) {
                    (qa2 as any)[tq] = tqa;
                  }
                }
              }
              fdData[q] = qa2;
            }
          }
          fd.data = fdData;
        }

        // Extract instance key and optional email — port of Python lines 197-209
        let fiEmail: string | null = null;
        let fi: string;
        if (fiRaw.includes('-')) {
          const fiArr = fiRaw.split('-');
          if (!fiArr[1].includes('@')) {
            fiEmail = fiArr[0];
            fi = fiArr[1];
          } else {
            fi = fiArr[0];
            fiEmail = fiArr[1];
          }
        } else {
          fi = fiRaw;
        }

        // Store in userDataByEmail
        if (!(ue in userDataByEmail)) userDataByEmail[ue] = {};
        if (!(ft in userDataByEmail[ue])) {
          userDataByEmail[ue][ft] = {};
          userDataByEmail[ue][ft][KEY] = {};
        }
        if (!(fi in userDataByEmail[ue][ft])) userDataByEmail[ue][ft][fi] = {};

        if (fiEmail) {
          userDataByEmail[ue][ft][fi][fiEmail] = fd;
        } else {
          userDataByEmail[ue][ft][fi] = fd;
        }
      }
    }
  }

  // [KEYRESET] Clear past matches — port of Python lines 226-230
  for (const e of Object.keys(userDataByEmail)) {
    if ('ttc_application' in userDataByEmail[e]) {
      if (!(KEY in userDataByEmail[e]['ttc_application'])) {
        userDataByEmail[e]['ttc_application'][KEY] = {};
      }
      userDataByEmail[e]['ttc_application'][KEY]['enrolled_matches'] = {};
      userDataByEmail[e]['ttc_application'][KEY]['org_course_matches'] = {};
    }
  }

  // Enrolled people matching (O(n²)) — port of Python lines 235-283
  for (const c1e of Object.keys(userDataByEmail)) {
    const c1TtcApp = userDataByEmail[c1e]['ttc_application'];
    if (!c1TtcApp) continue;

    for (const c1fi of getReportableInstanceKeys(c1TtcApp, [KEY])) {

      const c1 = c1TtcApp[c1fi];
      const c1d = c1?.data || {};

      const c1fn = (c1d.i_fname || '').trim().toLowerCase();
      const c1ln = (c1d.i_lname || '').trim().toLowerCase();
      const c1n = c1fn + ' ' + c1ln;

      for (const c2e of Object.keys(userDataByEmail)) {
        if (c1e === c2e) continue;

        const c2TtcApp = userDataByEmail[c2e]['ttc_application'];
        if (!c2TtcApp) continue;

        for (const c2fi of Object.keys(c2TtcApp)) {
          const c2 = c2TtcApp[c2fi];
          const c2d = c2?.data || {};

          const c2fn = (c2d.i_fname || '').trim().toLowerCase();
          const c2ln = (c2d.i_lname || '').trim().toLowerCase();
          const c2n = c2fn + ' ' + c2ln;

          if (c1n === c2n) continue;

          // Enrolled people matching
          const c1Enrolled = c1d.i_enrolled_people;
          const c2Enrolled = c2d.i_enrolled_people;

          if (Array.isArray(c1Enrolled) && Array.isArray(c2Enrolled)) {
            for (const c1en of c1Enrolled) {
              for (const c2en of c2Enrolled) {
                let isMatched = false;
                const ee1 = (c1en.i_enrollment_email || '').trim().toLowerCase();
                const ee2 = (c2en.i_enrollment_email || '').trim().toLowerCase();
                const en1 = (c1en.i_enrollment_name || '').trim().toLowerCase();
                const en2 = (c2en.i_enrollment_name || '').trim().toLowerCase();
                const ec1 = (c1en.i_enrollment_city || '').trim().toLowerCase();
                const ec2 = (c2en.i_enrollment_city || '').trim().toLowerCase();
                const es1 = (c1en.i_enrollment_state || '').trim().toLowerCase();
                const es2 = (c2en.i_enrollment_state || '').trim().toLowerCase();

                if (ee1.includes('@') && ee1 === ee2) isMatched = true;
                if (en1 === en2 && ec1 === ec2 && es1 === es2) isMatched = true;

                if (isMatched) {
                  if (!(KEY in userDataByEmail[c1e]['ttc_application'])) {
                    userDataByEmail[c1e]['ttc_application'][KEY] = {};
                  }
                  if (!('enrolled_matches' in userDataByEmail[c1e]['ttc_application'][KEY])) {
                    userDataByEmail[c1e]['ttc_application'][KEY]['enrolled_matches'] = {};
                  }
                  const matches = userDataByEmail[c1e]['ttc_application'][KEY]['enrolled_matches'];
                  if (!(c2e in matches)) {
                    matches[c2e] = new Set<string>();
                  }
                  matches[c2e].add(`${en2} <${ee2}>`);
                }
              }
            }
          }

          // Org course matching — port of Python lines 285-337
          const c1OrgCourses = c1d.i_org_courses;
          const c2OrgCourses = c2d.i_org_courses;

          if (Array.isArray(c1OrgCourses) && Array.isArray(c2OrgCourses)) {
            for (const c1oc of c1OrgCourses) {
              const ofd1 = (c1oc.i_org_course_from_date || '').trim();
              const otd1 = (c1oc.i_org_course_to_date || '').trim();
              const oc1 = (c1oc.i_org_course_city || '').trim().toLowerCase();
              const os1 = (c1oc.i_org_course_state || '').trim().toLowerCase();
              const olt1 = (c1oc.i_org_course_leadteacher || '').trim().toLowerCase();

              const re1 = olt1.match(/^[^a-z]{0,3}([a-z]+ [a-z]+)/i);
              const oltn1 = re1 ? re1[1] : '';

              for (const c2oc of c2OrgCourses) {
                let isMatched = false;
                const ofd2 = (c2oc.i_org_course_from_date || '').trim();
                const otd2 = (c2oc.i_org_course_to_date || '').trim();
                const oc2 = (c2oc.i_org_course_city || '').trim().toLowerCase();
                const os2 = (c2oc.i_org_course_state || '').trim().toLowerCase();
                const olt2 = (c2oc.i_org_course_leadteacher || '').trim().toLowerCase();

                const re2 = olt2.match(/^[^a-z]{0,3}([a-z]+ [a-z]+)/i);
                const oltn2 = re2 ? re2[1] : '';

                if (
                  (ofd1 === ofd2 || otd1 === otd2) &&
                  oc1 === oc2 &&
                  os1 === os2 &&
                  oltn1 &&
                  oltn2 &&
                  (oltn1 === oltn2 || olt1.includes(oltn2) || olt2.includes(oltn1))
                ) {
                  isMatched = true;
                }

                if (isMatched) {
                  if (!(KEY in userDataByEmail[c1e]['ttc_application'])) {
                    userDataByEmail[c1e]['ttc_application'][KEY] = {};
                  }
                  if (!('org_course_matches' in userDataByEmail[c1e]['ttc_application'][KEY])) {
                    userDataByEmail[c1e]['ttc_application'][KEY]['org_course_matches'] = {};
                  }
                  const matches = userDataByEmail[c1e]['ttc_application'][KEY]['org_course_matches'];
                  if (!(c2e in matches)) {
                    matches[c2e] = new Set<string>();
                  }
                  matches[c2e].add(`${ofd2} - ${otd2} (${oc2}, ${os2} - ${oltn2})`);
                }
              }
            }
          }
        }
      }
    }
  }

  // Convert Sets to Arrays for JSON serialization
  convertSetsToArrays(userDataByEmail);

  await writeJson(GCS_PATHS.USER_INTEGRITY_BY_USER, userDataByEmail);
}

/**
 * Port of Integrity.post_load_user_integrity() (Python lines 359-400).
 * Generates CSV of applicants and their enrolled people.
 */
export async function postLoadUserIntegrity(): Promise<void> {
  let userDataByEmail: any = {};
  try {
    userDataByEmail = await readJson(GCS_PATHS.USER_INTEGRITY_BY_USER);
  } catch {
    // File doesn't exist
  }

  let enrolledCSV = 'Applicant Name,Applicant Email,Enrolled Name,Enrolled Email\n';

  for (const c1e of Object.keys(userDataByEmail)) {
    const ttcApp = userDataByEmail[c1e]['ttc_application'];
    if (!ttcApp) continue;

    for (const c1fi of getReportableInstanceKeys(ttcApp, [KEY])) {

      const c1 = ttcApp[c1fi];
      const c1d = c1?.data || {};

      const c1fn = (c1d.i_fname || '').trim().toLowerCase();
      const c1ln = (c1d.i_lname || '').trim().toLowerCase();
      const c1n = c1fn + ' ' + c1ln;

      enrolledCSV += `${c1n},${c1e},,\n`;

      const enrolledPeople = c1d.i_enrolled_people;
      if (Array.isArray(enrolledPeople)) {
        for (const c1en of enrolledPeople) {
          const ee1 = (c1en.i_enrollment_email || '').trim().toLowerCase();
          const en1 = (c1en.i_enrollment_name || '').trim().toLowerCase();
          enrolledCSV += `,,${en1},${ee1}\n`;
        }
      }
    }
  }

  // Normalize unicode to ASCII — port of unicodedata.normalize('NFKD', ...).encode('ascii', 'ignore')
  enrolledCSV = enrolledCSV.normalize('NFKD').replace(/[\u0300-\u036f]/g, '');

  await writeText(GCS_PATHS.APPLICANT_ENROLLED_LIST, enrolledCSV);
}

/**
 * Port of Integrity.get_user_integrity_by_user() (Python lines 74-78).
 * Reads the integrity JSON from GCS.
 */
export async function getUserIntegrityByUser(): Promise<unknown> {
  return readJson(GCS_PATHS.USER_INTEGRITY_BY_USER);
}
